File: lib/search-service.ts
import type { Album } from './types';
import { sanitizeAlbum } from './security';

// Simple in-memory cache for search results
const searchCache = new Map<string, { data: Album[], timestamp: number }>();
const CACHE_TTL = 5 * 60 * 1000; // 5 minutes
const MAX_CACHE_SIZE = 50;

const TRUSTED_JSONP_DOMAINS = ['api.deezer.com', 'itunes.apple.com'];

/**
 * Wraps a search function with a simple in-memory cache.
 * Improves performance for repeated searches and reduces network requests.
 */
function withCache<T extends any[]>(
  provider: string,
  fn: (query: string, ...args: T) => Promise<Album[]>
) {
  return async (query: string, ...args: T): Promise<Album[]> => {
    const trimmedQuery = query.trim().toLowerCase();
    if (!trimmedQuery) return [];

    // Note: token is intentionally omitted from the cache key as search results
    // for the same query are expected to be identical across valid tokens.
    const cacheKey = `${provider}:${trimmedQuery}`;
    const cached = searchCache.get(cacheKey);
    const now = Date.now();

    if (cached && now - cached.timestamp < CACHE_TTL) {
      return cached.data;
    }

    const data = await fn(query, ...args);

    // Maintain cache size
    if (searchCache.size >= MAX_CACHE_SIZE) {
      const oldestKey = searchCache.keys().next().value;
      if (oldestKey !== undefined) {
        searchCache.delete(oldestKey);
      }
    }

    searchCache.set(cacheKey, { data, timestamp: now });
    return data;
  };
}

// Helper for JSONP requests to Deezer and Apple APIs
function jsonp<T>(url: string): Promise<T> {
  // Domain whitelist check for defense-in-depth
  try {
    const parsed = new URL(url);
    if (!TRUSTED_JSONP_DOMAINS.includes(parsed.hostname)) {
      throw new Error(`Untrusted JSONP domain: ${parsed.hostname}`);
    }
  } catch (e) {
    return Promise.reject(e instanceof Error ? e : new Error('Invalid JSONP URL'));
  }

  return new Promise((resolve, reject) => {
    const randomArray = new Uint32Array(1);
    crypto.getRandomValues(randomArray);
    const callbackName = `jsonp_callback_${randomArray[0]}`;
    const script = document.createElement('script');

    // @ts-ignore
    window[callbackName] = (data: T) => {
      delete (window as any)[callbackName];
      document.body.removeChild(script);
      resolve(data);
    };

    script.src = `${url}${url.indexOf('?') >= 0 ? '&' : '?'}callback=${callbackName}`;
    script.onerror = () => {
      delete (window as any)[callbackName];
      document.body.removeChild(script);
      reject(new Error(`JSONP request failed for ${url}`));
    };
    document.body.appendChild(script);
  });
}

async function searchAlbumsDeezerInternal(query: string): Promise<Album[]> {
  try {
    const data = await jsonp<any>(
      `https://api.deezer.com/search/album?q=${encodeURIComponent(query)}&limit=20&output=jsonp`
    );

    if (data.error) {
      throw new Error(data.error.message || 'Deezer search error');
    }

    return data.data.map((item: any) => sanitizeAlbum({
      id: `deezer-${item.id}`,
      name: item.title,
      artist: item.artist.name,
      imageUrl: item.cover_big || item.cover_xl || item.cover_medium,
      releaseDate: '', // Deezer search API doesn't provide release date at top level
      totalTracks: item.nb_tracks,
      externalUrl: item.link,
    }));
  } catch (error) {
    console.error('Deezer search error:', error);
    throw error;
  }
}

async function searchAlbumsSpotifyInternal(query: string, token: string | null): Promise<Album[]> {
  if (!token) return [];

  try {
    const response = await fetch(
      `https://api.spotify.com/v1/search?q=${encodeURIComponent(query)}&type=album&limit=20`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );

    if (!response.ok) {
      if (response.status === 401) {
        throw new Error('Spotify session expired. Please reconnect.');
      }
      throw new Error('Spotify search failed');
    }

    const data = await response.json();

    return data.albums.items.map((item: any) => sanitizeAlbum({
      id: `spotify-${item.id}`,
      spotifyId: item.id,
      name: item.name,
      artist: item.artists.map((a: any) => a.name).join(', '),
      imageUrl: item.images[0]?.url || '/placeholder.svg',
      releaseDate: item.release_date,
      totalTracks: item.total_tracks,
      externalUrl: item.external_urls.spotify,
      spotifyUrl: item.external_urls.spotify,
    }));
  } catch (error) {
    console.error('Spotify search error:', error);
    throw error;
  }
}

async function searchAlbumsAppleInternal(query: string): Promise<Album[]> {
  try {
    const data = await jsonp<any>(
      `https://itunes.apple.com/search?term=${encodeURIComponent(query)}&entity=album&limit=20`
    );

    if (!data.results) {
      return [];
    }

    return data.results.map((item: any) => sanitizeAlbum({
      id: `apple-${item.collectionId}`,
      name: item.collectionName,
      artist: item.artistName,
      imageUrl: item.artworkUrl100.replace('100x100bb', '600x600bb'), // Get higher res image
      releaseDate: item.releaseDate,
      totalTracks: item.trackCount,
      externalUrl: item.collectionViewUrl,
    }));
  } catch (error) {
    console.error('Apple search error:', error);
    throw error;
  }
}

export const searchAlbumsDeezer = withCache('deezer', searchAlbumsDeezerInternal);
export const searchAlbumsApple = withCache('apple', searchAlbumsAppleInternal);
export const searchAlbumsSpotify = withCache('spotify', searchAlbumsSpotifyInternal);

File: internal/service/search_service.go
package service

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

type SearchResult struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Artist      string `json:"artist"`
	ImageUrl    string `json:"imageUrl"`
	ExternalUrl string `json:"externalUrl"`
}

type SearchService struct {
	client       *http.Client
	spotifyID     string
	spotifySecret string
	spotifyToken  string
}

func NewSearchService(spotifyID, spotifySecret string) *SearchService {
	return &SearchService{
		client:        &http.Client{},
		spotifyID:     spotifyID,
		spotifySecret: spotifySecret,
	}
}

func (s *SearchService) SearchApple(query string) ([]SearchResult, error) {
	resp, err := s.client.Get(fmt.Sprintf("https://itunes.apple.com/search?term=%s&entity=album&limit=25", url.QueryEscape(query)))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var data struct {
		Results []struct {
			CollectionId   int    `json:"collectionId"`
			CollectionName string `json:"collectionName"`
			ArtistName     string `json:"artistName"`
			ArtworkUrl100  string `json:"artworkUrl100"`
			CollectionViewUrl string `json:"collectionViewUrl"`
		} `json:"results"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}

	results := make([]SearchResult, 0, len(data.Results))
	for _, r := range data.Results {
		results = append(results, SearchResult{
			ID:          fmt.Sprintf("apple-%d", r.CollectionId),
			Name:        r.CollectionName,
			Artist:      r.ArtistName,
			ImageUrl:    r.ArtworkUrl100,
			ExternalUrl: r.CollectionViewUrl,
		})
	}
	return results, nil
}

func (s *SearchService) SearchSpotify(query string) ([]SearchResult, error) {
	if s.spotifyID == "" || s.spotifySecret == "" {
		return nil, fmt.Errorf("spotify credentials not configured")
	}

	if s.spotifyToken == "" {
		if err := s.refreshSpotifyToken(); err != nil {
			return nil, err
		}
	}

	req, _ := http.NewRequest("GET", fmt.Sprintf("https://api.spotify.com/v1/search?q=%s&type=album&limit=25", url.QueryEscape(query)), nil)
	req.Header.Set("Authorization", "Bearer "+s.spotifyToken)

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusUnauthorized {
		if err := s.refreshSpotifyToken(); err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "Bearer "+s.spotifyToken)
		resp, err = s.client.Do(req)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()
	}

	var data struct {
		Albums struct {
			Items []struct {
				ID      string `json:"id"`
				Name    string `json:"name"`
				Artists []struct {
					Name string `json:"name"`
				} `json:"artists"`
				Images []struct {
					URL string `json:"url"`
				} `json:"images"`
				ExternalUrls struct {
					Spotify string `json:"spotify"`
				} `json:"external_urls"`
			} `json:"items"`
		} `json:"albums"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}

	results := make([]SearchResult, 0, len(data.Albums.Items))
	for _, r := range data.Albums.Items {
		artist := ""
		if len(r.Artists) > 0 {
			artist = r.Artists[0].Name
		}
		imageUrl := ""
		if len(r.Images) > 0 {
			imageUrl = r.Images[0].URL
		}
		results = append(results, SearchResult{
			ID:          "spotify-" + r.ID,
			Name:        r.Name,
			Artist:      artist,
			ImageUrl:    imageUrl,
			ExternalUrl: r.ExternalUrls.Spotify,
		})
	}
	return results, nil
}

func (s *SearchService) refreshSpotifyToken() error {
	data := url.Values{}
	data.Set("grant_type", "client_credentials")

	req, _ := http.NewRequest("POST", "https://accounts.spotify.com/api/token", strings.NewReader(data.Encode()))
	req.SetBasicAuth(s.spotifyID, s.spotifySecret)
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := s.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	var res struct {
		AccessToken string `json:"access_token"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return err
	}

	s.spotifyToken = res.AccessToken
	return nil
}

func (s *SearchService) SearchDeezer(query string) ([]SearchResult, error) {
	resp, err := s.client.Get(fmt.Sprintf("https://api.deezer.com/search/album?q=%s&limit=25", url.QueryEscape(query)))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var data struct {
		Data []struct {
			ID    int    `json:"id"`
			Title string `json:"title"`
			Artist struct {
				Name string `json:"name"`
			} `json:"artist"`
			CoverMedium string `json:"cover_medium"`
			Link        string `json:"link"`
		} `json:"data"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}

	results := make([]SearchResult, 0, len(data.Data))
	for _, r := range data.Data {
		results = append(results, SearchResult{
			ID:          fmt.Sprintf("deezer-%d", r.ID),
			Name:        r.Title,
			Artist:      r.Artist.Name,
			ImageUrl:    r.CoverMedium,
			ExternalUrl: r.Link,
		})
	}
	return results, nil
}
